### PR TITLE
Allow "0" prefetch value in `concatMap`

### PR DESCRIPTION
**Pull request by bsideup**
`concatMap` can be very helpful in combination with `window(n)`.

But the current implementation enforces the mandatory prefetch,
making it request another window without waiting for the completion of the inner `Publisher`.

Given `concatMap`'s nature, we can make it accept `0` prefetch value,
so that it will request an item on a first request, and next request will be
on inner `Publisher`'s completion.

---

## Implementation Details

As an experiment, I decided to try the state-based approach when implementing it.
The end result was benchmarked against [more "classic" implementation](https://gist.github.com/OlegDokuka/a48911846349d381d7bb0437d74decc5) (by @OlegDokuka).

The benchmark results how that this approach performs well and sometimes even outperforms:
```text

Benchmark                                                     Mode  Cnt         Score        Error  Units
FluxConcatMapNoPrefetchBenchmark.state                      sample  304  50641650.526 ± 662739.360  ns/op
FluxConcatMapNoPrefetchBenchmark.state:state·p0.00          sample       46727168.000               ns/op
FluxConcatMapNoPrefetchBenchmark.state:state·p0.50          sample       49676288.000               ns/op
FluxConcatMapNoPrefetchBenchmark.state:state·p0.90          sample       54198272.000               ns/op
FluxConcatMapNoPrefetchBenchmark.state:state·p0.95          sample       59097088.000               ns/op
FluxConcatMapNoPrefetchBenchmark.state:state·p0.99          sample       65745715.200               ns/op
FluxConcatMapNoPrefetchBenchmark.state:state·p0.999         sample       66519040.000               ns/op
FluxConcatMapNoPrefetchBenchmark.state:state·p0.9999        sample       66519040.000               ns/op
FluxConcatMapNoPrefetchBenchmark.state:state·p1.00          sample       66519040.000               ns/op
FluxConcatMapNoPrefetchBenchmark.zclassic                   sample  300  51276021.760 ± 547219.851  ns/op
FluxConcatMapNoPrefetchBenchmark.zclassic:zclassic·p0.00    sample       47316992.000               ns/op
FluxConcatMapNoPrefetchBenchmark.zclassic:zclassic·p0.50    sample       50462720.000               ns/op
FluxConcatMapNoPrefetchBenchmark.zclassic:zclassic·p0.90    sample       54899507.200               ns/op
FluxConcatMapNoPrefetchBenchmark.zclassic:zclassic·p0.95    sample       57668403.200               ns/op
FluxConcatMapNoPrefetchBenchmark.zclassic:zclassic·p0.99    sample       63562055.680               ns/op
FluxConcatMapNoPrefetchBenchmark.zclassic:zclassic·p0.999   sample       64487424.000               ns/op
FluxConcatMapNoPrefetchBenchmark.zclassic:zclassic·p0.9999  sample       64487424.000               ns/op
FluxConcatMapNoPrefetchBenchmark.zclassic:zclassic·p1.00    sample       64487424.000               ns/op
```

## Stress testing

To verify the implementation, I added [JCstress](https://openjdk.java.net/projects/code-tools/jcstress/)-based tests.
JCstress is a great framework for testing race conditions, and it is very easy to use (as seen in the implemented tests).

### CODE REVIEW

File: reactor-test/build.gradle [hunk 1]
```diff
@@ -51,32 +51,30 @@ dependencies {
 	testCompile "org.mockito:mockito-core:$mockitoVersion"
 }
 
-task downloadBaseline {
+task downloadBaseline(type: Download) {
```
Comment (simonbasle) [inline reactor-test/build.gradle:54]: can you make it part of a separate commit and prepare the PR to be rebase-merged ?

File: reactor-core/src/main/java/reactor/core/publisher/FluxConcatMapNoPrefetch.java [hunk 2]
```diff
@@ -0,0 +1,299 @@
...
+					default:
+						Operators.onErrorDropped(e, currentContext());
+						return;
+				}
+			}
+		}
+
+		@Override
+		public void request(long n) {
+			if (STATE.compareAndSet(this, State.INITIAL, State.REQUESTED)) {
```
Comment (simonbasle) [inline reactor-core/src/main/java/reactor/core/publisher/FluxConcatMapNoPrefetch.java:279]: here I think it is worth it detailing the behavior in a comment:
 - the fact that obviously the only time a downstream request directly results in an `upstream` request is the first time
 - detail of what happens when `request(10)` is done but the inner only serves eg. 3 elements (3 inner next will be tracked as produced by the `ConcatMapInner`, then inner onComplete will update the `requested` tracker, then `innerComplete` will transition to REQUESTED state and request one more source, swap the subscription and request the 7 remaining demand)

Comment (simonbasle) [reply reactor-core/src/main/java/reactor/core/publisher/FluxConcatMapNoPrefetch.java:279]: (this is absolutely not necessary, but after reviewing that was the one spot I had to double back over)

File: reactor-core/src/test/java/reactor/core/publisher/FluxConcatMapTest.java [hunk 3]
```diff
@@ -1226,4 +1227,46 @@ public void discardDelayedOnDrainMapperError() {
 		            .verifyThenAssertThat()
 		            .hasDiscardedExactly(1);
 	}
+
+	@Test
+	public void noRequestBeforeOnCompleteWithZeroPrefetch() {
```
Comment (simonbasle) [inline reactor-core/src/test/java/reactor/core/publisher/FluxConcatMapTest.java:1232]: would it make sense to have additional tests for the concrete class in a separate test case, on top of that one (which tests the API goes down the `FluxConcatMapNoPrefetch` path I guess) and the stress test ? There might not be much left to test other than the `Scannable`, that said, but still... this test can actually be done in the separate test class as well.

File: reactor-core/src/main/java/reactor/core/publisher/FluxConcatMapNoPrefetch.java [hunk 4]
```diff
@@ -0,0 +1,299 @@
...
+				default:
+					Operators.onDiscard(value, currentContext());
+					break;
+			}
+		}
+
+		@Override
+		public void innerComplete() {
+			for (State previousState = this.state; ; previousState = this.state) {
+				switch (previousState) {
```
[the remark this answers is not in the document]
Comment (simonbasle) [reply reactor-core/src/main/java/reactor/core/publisher/FluxConcatMapNoPrefetch.java:238]: (this generally applies to CAS in that class)

[the remark this answers is not in the document]
Comment (bsideup) [reply reactor-core/src/main/java/reactor/core/publisher/FluxConcatMapNoPrefetch.java:238]: I thought about doing it, but then realized that it is not copy/paste friendly and we may slip some bugs due to it.
When reviewing, having to check `previousState` IMO is actually better, because it forces you to double check that we're in the correct `case` branch. Otherwise imagine accidentally approving something like:
```java
switch (previousState) {
	case ACTIVE:
		if (!STATE.compareAndSet(this, ACTIVE, State.REQUESTED)) {
			continue;
		}
		upstream.request(1);
		return;
	case LAST_ACTIVE:
		if (!STATE.compareAndSet(this, ACTIVE, State.TERMINATED)) {
			continue;
		}
		actual.onComplete();
		return;
	default:
		return;
}
```

if not covered with tests (hehe :D), we will ship a potentially infinite loop to production, and the cost of such mistake would be high

File: reactor-core/src/test/java/reactor/core/publisher/FluxConcatMapTest.java (hunk 3 above)
Comment (bsideup) [reply reactor-core/src/test/java/reactor/core/publisher/FluxConcatMapTest.java:1232]: I actually ran (locally) `FluxConcatMapTest` with `*NoPrefetch` (by adding `prefetch == 0 || true` to `Flux.java`) and it passed all but the prefetch ones.

Glad that you asked, because I wanted to discuss the testing approach here. We could reuse the same set of tests by replacing every `.concatMap` with `.as(this.concatMap(...))`. This will be a massive change in this PR but IMO worth it. WDYT?

**Comment by bsideup** [comment]
@simonbasle I will submit the japicmp change separately, since it can be considered a general improvement

File: reactor-core/src/main/java/reactor/core/publisher/Flux.java [hunk 5]
```diff
@@ -3663,13 +3663,16 @@ public final T blockLast(Duration timeout) {
...
 	 * @param mapper the function to transform this sequence of T into concatenated sequences of V
-	 * @param prefetch the inner source produced demand
+	 * @param prefetch the inner source produced demand (set it to 0 if you don't want it to prefetch)
 	 * @param <V> the produced concatenated type
 	 *
 	 * @return a concatenated {@link Flux}
 	 */
 	public final <V> Flux<V> concatMap(Function<? super T, ? extends Publisher<? extends V>>
 			mapper, int prefetch) {
+		if (prefetch == 0) {
```
Comment (simonbasle) [inline reactor-core/src/main/java/reactor/core/publisher/Flux.java:3673]: wait there's actually an inconsistency. your implementation seem to always delay errors, so it should be an overload of `concatMapDelayErrors`

File: reactor-core/src/main/java/reactor/core/publisher/FluxConcatMapNoPrefetch.java [hunk 6]
```diff
@@ -0,0 +1,298 @@
...
+			Flux<? extends T> source,
+			Function<? super T, ? extends Publisher<? extends R>> mapper
+	) {
+		super(source);
+		this.mapper = Objects.requireNonNull(mapper, "mapper");
+	}
+
+	@Override
+	public CoreSubscriber<? super T> subscribeOrReturn(CoreSubscriber<? super R> actual) {
+		if (FluxFlatMap.trySubscribeScalarMap(source, actual, mapper, false, true)) {
```
Comment (simonbasle) [inline reactor-core/src/main/java/reactor/core/publisher/FluxConcatMapNoPrefetch.java:55]: see comment on `Flux` relative to that second boolean (`expectDelayErrors`) being `true`.

File: reactor-core/src/main/java/reactor/core/publisher/FluxConcatMapNoPrefetch.java (hunk 4 above)
[the remark this answers is not in the document]
Comment (simonbasle) [reply reactor-core/src/main/java/reactor/core/publisher/FluxConcatMapNoPrefetch.java:237]: ok, that makes sense as well. let's keep it that way then.

File: reactor-core/src/test/java/reactor/core/publisher/FluxConcatMapTest.java (hunk 3 above)
Comment (simonbasle) [reply reactor-core/src/test/java/reactor/core/publisher/FluxConcatMapTest.java:1233]: are you talking about mutualizing the test code, or duplicating the relevant tests? I would definitely agree on duplicating, but if reusing I'm expecting only a subset of the tests make sense for `FluxConcatMapNoPrefetch`?

Comment (bsideup) [reply reactor-core/src/test/java/reactor/core/publisher/FluxConcatMapTest.java:1233]: or, actually, an easier way would be to make the prefetch in the test explicit... yes, I think that's the way to go :) Will do.

File: reactor-core/src/main/java/reactor/core/publisher/Flux.java (hunk 5 above)
Comment (bsideup) [reply reactor-core/src/main/java/reactor/core/publisher/Flux.java:3673]: Oh... nice catch! Thanks!

File: reactor-core/src/main/java/reactor/core/publisher/FluxConcatMapNoPrefetch.java [hunk 7]
```diff
@@ -0,0 +1,298 @@
...
+			return null;
+		}
+
+		return new FluxConcatMapNoPrefetchSubscriber<>(actual, mapper);
+	}
+
+	static final class FluxConcatMapNoPrefetchSubscriber<T, R> implements FluxConcatMapSupport<T, R> {
+
+		enum State {
```
Comment (ericbottard) [inline reactor-core/src/main/java/reactor/core/publisher/FluxConcatMapNoPrefetch.java:64]: Could you document the transition diagram, or at least explain the states in more detail (in particular LAST_ACTIVE)?

Comment (bsideup) [reply reactor-core/src/main/java/reactor/core/publisher/FluxConcatMapNoPrefetch.java:64]: Unless you have a good idea how to _code_ it, I don't think doing a diagram is a good idea, since it may easily get outdated :(

Comment (ericbottard) [reply reactor-core/src/main/java/reactor/core/publisher/FluxConcatMapNoPrefetch.java:69]: I meant at least documenting what transitions make sense, and what each (non trivial) state represents. That said, I would argue that if we were to produce a diagram for comprehension's sake and it got outdated, there is indeed a problem :)

Comment (OlegDokuka) [reply reactor-core/src/main/java/reactor/core/publisher/FluxConcatMapNoPrefetch.java:69]: I believe that code is the best documentation especially when state transition is not that complicated and can be easily derived from the enums.

File: reactor-core/src/main/java/reactor/core/publisher/Flux.java (hunk 5 above)
Comment (bsideup) [reply reactor-core/src/main/java/reactor/core/publisher/Flux.java:3664]: FYI should be fixed now

File: reactor-core/src/main/java/reactor/core/publisher/Flux.java [hunk 8]
```diff
@@ -472,10 +472,7 @@
 	 * @return a new {@link Flux} concatenating all inner sources sequences
 	 */
 	public static <T> Flux<T> concat(Publisher<? extends Publisher<? extends T>> sources, int prefetch) {
-		return onAssembly(new FluxConcatMap<>(from(sources),
-				identityFunction(),
-				Queues.get(prefetch), prefetch,
-				FluxConcatMap.ErrorMode.IMMEDIATE));
+		return wrap(sources).concatMap(identityFunction(), prefetch);
```
Comment (simonbasle) [inline reactor-core/src/main/java/reactor/core/publisher/Flux.java:475]: 👍, please just mention this change in the body of the commit message

Comment (bsideup) [reply reactor-core/src/main/java/reactor/core/publisher/Flux.java:475]: Isn't it just another way of writing the same and completely "internal"? Or did I change the behaviour? 😱

File: reactor-core/src/main/java/reactor/core/publisher/FluxConcatMapNoPrefetch.java (hunk 6 above)
Comment (bsideup) [reply reactor-core/src/main/java/reactor/core/publisher/FluxConcatMapNoPrefetch.java:60]: Seems to be hardcoded `true` here as well:
https://github.com/reactor/reactor-core/blob/cb74c0c377599723f4a51549f042040a555aa9d4/reactor-core/src/main/java/reactor/core/publisher/FluxConcatMap.java#L116

Since the source is Scalar, I guess it doesn't make a difference?

Comment (simonbasle) [reply reactor-core/src/main/java/reactor/core/publisher/FluxConcatMapNoPrefetch.java:60]: good point about the scalar nature 👍

File: reactor-core/src/main/java/reactor/core/publisher/Flux.java (hunk 8 above)
Comment (simonbasle) [reply reactor-core/src/main/java/reactor/core/publisher/Flux.java:475]: well, the `wrap` won't apply assembly hooks, but the `concatMap` will, so I think this is ok. it's just that the direct application of assembly hooks would have pointed tracebacks maybe a tiny bit more directly to `concat` (rather than having `concatMap` in the traceback), but I think that is ok.

Comment (bsideup) [reply reactor-core/src/main/java/reactor/core/publisher/Flux.java:475]: Oh, you're right! Actually, let's use `from` as it was before to avoid any side effects from this change. I will submit a separate PR that will use `wrap` where applicable and evaluate the impact without the rest of the concatMap changes, ok?

Comment (simonbasle) [reply reactor-core/src/main/java/reactor/core/publisher/Flux.java:475]: sounds good to me
